File: javascript/binary_gap.js
/*
 * Result for lesson:
 * https://codility.com/programmers/lessons/1-iterations/binary_gap/
 */

function solution(A) {
    var binary = A.toString(2);
    var binaryArray = binary.split("");
    var count = result = 0;
    var start=false;

    for(i=0; i<binaryArray.length; i++){
    	if(binaryArray[i]==="1"){
    		if(count > result){
    			result = count;
    			count = 0;
    		}
    		if(start){
    			count = 0;
    		}else{
    			start = true;
    		}
    	}
    	if(start && binaryArray[i]==="0"){
    		count += 1;
    	}
    }
    return result;
}

/*
 * tests
 */

function test(functionToTest, args, expected) {
	var result = functionToTest.apply(this, args);
    if (result !== expected) {
        console.log(args.toString(2));
        console.log("Expected " + expected + ", but was " + result);
    }
}

test(solution, [9], 2);
test(solution, [529], 4);
test(solution, [20], 1);
test(solution, [15], 0);
test(solution, [1041], 5);
test(solution, [0], 0);
test(solution, [1], 0);
test(solution, [66561], 9);
test(solution, [74901729], 4);

File: misc/BinaryGap.py
def solution(N):
    binary_value = bin(N)[2:]
    count = result = 0
    start = False
    
    for i in range(0,len(binary_value)):
        if binary_value[i] == '1':
            if count > result:
                result = count
                count = 0
            if start:
                count = 0
            else:
                start = True
        if start and (binary_value[i] == '0'):
            count = count + 1
    
    return result

def test(args, expect):
	print("Expect: " + str(expect) + ", was: " + str(solution(args)))

test(1041,5)
test(15,0)
test(10,1)
test(328, 2)